### PR TITLE
Work on new DiffBasedAnomalyDetector with different CV methodology + shuffling

### PR DESCRIPTION
**Pull request by hugobettmach**
* Add new subclass of `DiffBasedAnomalyDetector` called `DiffBasedKFCVAnomalyDetector` which:
  * Built to support threshold calculation when using a random k-fold cross-validation
    * Threshold calculated using a percentile of the validation errors calculated from the cross-validation process.
* Some improvements on original `DiffBasedAnomalyDetector`
  * Adds the possibility to shuffle data before fitting to avoid keeping the last `validation_split` % of the data as validation, and therefore never actually training in the most recent data.
  * Adds `smoothing_method` arg which defines which smoothing algorithm is used, for now possible to select from simple moving median, simple moving average or exponential weighted moving average.
* Fixes deprecated imports across the repository.

Examples:
1. How to call the `DiffBasedAnomalyDetector` with the new args:
```yaml
- dataset:
        row_filter: (`...` > 1500.0)
        tags:
        - ...
        - ...
        train_end_date: '2020-08-25T00:00:00+00:00'
        train_start_date: '2020-08-15T00:00:00+00:00'
      evaluation:
        cv:
          sklearn.model_selection.TimeSeriesSplit:
            n_splits: 3
        scoring_scaler: sklearn.preprocessing.MinMaxScaler
        seed: 0
      metadata:
        model-hierarchy: '...'
      model:
        gordo.machine.model.anomaly.diff.DiffBasedAnomalyDetector:
          base_estimator:
            sklearn.compose.TransformedTargetRegressor:
              regressor:
                sklearn.pipeline.Pipeline:
                  steps:
                  - sklearn.preprocessing.MinMaxScaler
                  - gordo.machine.model.models.KerasAutoEncoder:
                      batch_size: 128
                      callbacks:
                      - tensorflow.keras.callbacks.EarlyStopping:
                          monitor: val_loss
                          patience: 10
                          restore_best_weights: true
                      compression_factor: 0.5
                      encoding_layers: 1
                      epochs: 1000
                      func: tanh
                      kind: feedforward_hourglass
                      loss: mse
                      optimizer: Adam
                      out_func: linear
                      validation_split: 0.1
              transformer: sklearn.preprocessing.MinMaxScaler
          scaler: sklearn.preprocessing.MinMaxScaler
          shuffle: true
          smoothing_method: smm
          window: 144
```

2. Example of how to call the new subclass `DiffBasedKFCVAnomalyDetector`. Please note the cross-validation object should be passed under `evaluation`:
```yaml
- dataset:
        row_filter: (`...` > 1500.0)
        tags:
        - ...
        - ...
        train_end_date: '2020-08-25T00:00:00+00:00'
        train_start_date: '2020-08-15T00:00:00+00:00'
      evaluation:
        cv:
          sklearn.model_selection.KFold:
            n_splits: 5
            random_state: 0
            shuffle: true
        scoring_scaler: sklearn.preprocessing.MinMaxScaler
        seed: 0
      metadata:
        model-hierarchy: '...'
      model:
        gordo.machine.model.anomaly.diff.DiffBasedKFCVAnomalyDetector:
          base_estimator:
            sklearn.compose.TransformedTargetRegressor:
              regressor:
                sklearn.pipeline.Pipeline:
                  steps:
                  - sklearn.preprocessing.MinMaxScaler
                  - gordo.machine.model.models.KerasAutoEncoder:
                      batch_size: 128
                      callbacks:
                      - tensorflow.keras.callbacks.EarlyStopping:
                          monitor: val_loss
                          patience: 10
                          restore_best_weights: true
                      compression_factor: 0.5
                      encoding_layers: 1
                      epochs: 1000
                      func: tanh
                      kind: feedforward_hourglass
                      loss: mse
                      optimizer: Adam
                      out_func: linear
                      validation_split: 0.1
              transformer: sklearn.preprocessing.MinMaxScaler
          scaler: sklearn.preprocessing.MinMaxScaler
          shuffle: true
          smoothing_method: smm
          threshold_percentile: 0.99
          window: 144
```

### CODE REVIEW

**Review by henrikny — COMMENTED** [review]
This is a large PR which I think adds great value to Gordo.
It could easily be split up into at least two separate PRs, as reading the code changes gets a bit hard, everything considered.

Functionalising `smoothing_method`, adding shuffling and updating some dependencies is good!
I found it hard do read the complete scope of the testing for the new subclass.
Does it also evaluate actual data and asserts the returned results?

It has my approval, but I would like to see @koropets agree.

File: gordo/machine/model/anomaly/diff.py [hunk 1]
```diff
@@ -159,14 +208,16 @@ def cross_validate(
...
+        cv_output = c_val(self, X=X, y=y, **kwargs)
 
         self.feature_thresholds_per_fold_ = pd.DataFrame()
         self.aggregate_thresholds_per_fold_ = {}
         self.smooth_feature_thresholds_per_fold_ = pd.DataFrame()
         self.smooth_aggregate_thresholds_per_fold_ = {}
         smooth_aggregate_threshold_fold = None
         smooth_tag_thresholds_fold = None
+        if self.window is not None and self.smoothing_method is None:
+            self.smoothing_method = "smm"
```
[the remark this answers is not in the document]
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py]: Yes it could, my dilemma here is that defining a `smoothing_method` only makes sense if the `window` parameter exists, otherwise it should probably be set to `None`.
This was built like this to account for existing configs where smoothing method has not been defined since it wasn't available yet but `window` was defined.
If you strongly disagree I can indeed set it as default in the class definition.

File: gordo/machine/model/anomaly/diff.py [hunk 2]
```diff
@@ -231,24 +283,41 @@ def _scaled_mse_per_timestep(
...
         mse_per_time_step = ((scaled_y_pred - scaled_y_true) ** 2).mean(axis=1)
         return pd.Series(mse_per_time_step)
 
+    @staticmethod
+    def _absolute_error(
+        y_true: Union[pd.DataFrame, np.ndarray],
+        y_pred: Union[pd.DataFrame, np.ndarray],
+    ) -> pd.DataFrame:
+
+        return pd.DataFrame(np.abs(y_true - y_pred))
```
[the remark this answers is not in the document]
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py:310]: I guess this comes down to personal preference, I find it more elegant having that method, since it is used more than once in the code and with the method name, one knows immediately what it is doing.

File: gordo/machine/model/anomaly/diff.py [hunk 3]
```diff
@@ -183,9 +234,10 @@ def cross_validate(
             scaled_mse = self._scaled_mse_per_timestep(split_model, y_true, y_pred)
 
             # Absolute error
-            mae = pd.DataFrame(np.abs(y_pred - y_true))
+            mae = self._absolute_error(y_true, y_pred)
```
[the remark this answers is not in the document]
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py:236]: See reply above.

File: gordo/machine/model/anomaly/diff.py [hunk 4]
```diff
@@ -322,30 +391,31 @@ def anomaly(
         )
 
         if self.window is not None:
+            if self.smoothing_method is None:
+                self.smoothing_method = "smm"
```
[the remark this answers is not in the document]
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py]: See reply above.

File: gordo/machine/model/anomaly/diff.py [hunk 5]
```diff
@@ -398,8 +481,195 @@ def anomaly(
...
+            zip(kwargs["cv"].split(X, y), cv_output["estimator"])
+        ):
+            y_pred.iloc[test_idxs] = split_model.predict(
+                X.iloc[test_idxs].to_numpy()
+                if isinstance(X, pd.DataFrame)
+                else X[test_idxs]
+            )
+
+        # Calculate global threshold
+        self.aggregate_threshold_ = self._calculate_aggregate_threshold(self, y, y_pred)
```
[the remark this answers is not in the document]
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py]: Yes you are absolutely right, I don't. Will correct.

**Comment by hugobettmach** [comment]
> This is a large PR which I think adds great value to Gordo.
> It could easily be split up into at least two separate PRs, as reading the code changes gets a bit hard, everything considered.
> 
> Functionalising `smoothing_method`, adding shuffling and updating some dependencies is good!
> I found it hard do read the complete scope of the testing for the new subclass.
> Does it also evaluate actual data and asserts the returned results?
> 
> It has my approval, but I would like to see @koropets agree.

Thanks @henrikny .
Regarding your question about the testing, if you look at the test file I followed the same pattern as it was designed before. There is a test function specifically for this method `test_diff_kfcv_detector` and to test the other stuff (e.g. metadata, thresholds) I added the following condition or similar to the other tests so that all variations of the Diff classes are covered: `@pytest.mark.parametrize("mode", ("tscv", "kfcv"))`, where `tscv` refers to the original class and `kfcv` to the new subclass.

File: gordo/machine/model/anomaly/diff.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (henrikny) [reply gordo/machine/model/anomaly/diff.py]: I see. No worries :) I guess there could be a default `window` parameter as well.

File: gordo/machine/model/anomaly/diff.py [hunk 6]
```diff
@@ -159,14 +208,16 @@ def cross_validate(
...
-        cv_output = cross_validate(self, X=X, y=y, **kwargs)
+        cv_output = c_val(self, X=X, y=y, **kwargs)
 
         self.feature_thresholds_per_fold_ = pd.DataFrame()
         self.aggregate_thresholds_per_fold_ = {}
         self.smooth_feature_thresholds_per_fold_ = pd.DataFrame()
         self.smooth_aggregate_thresholds_per_fold_ = {}
         smooth_aggregate_threshold_fold = None
         smooth_tag_thresholds_fold = None
+        if self.window is not None and self.smoothing_method is None:
```
Comment (koropets) [inline gordo/machine/model/anomaly/diff.py]: I think these two lines should be moved into `__init__` method, because in this way `self.smoothing_method` stay `None` until `cross_validation()` method be called

File: gordo/machine/model/anomaly/diff.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py]: Following @koropets suggestion I moved the whole `if` statement it to `__init__()`, which is the best compromise I guess, making sure `smoothing_method` is set if `window` exists when creating a new class instance, while not being a default. I guess this solves this?

File: gordo/machine/model/anomaly/diff.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py]: Same as above, when placing this condition in `__init__()` it was not needed here anymore and has been removed.

File: gordo/machine/model/anomaly/diff.py (hunk 6 above)
Comment (hugobettmach) [reply gordo/machine/model/anomaly/diff.py]: Thanks, good idea. Moved the whole `if` statement it to `__init__()`, agree this is the best solution. Also had to adjust tests accordingly.
